### PR TITLE
Release: migrate release process to release-it

### PR DESCRIPTION
**Pull request by timmywil**
### Summary ###

- This migrates the release process from jquery-release to [release-it](https://github.com/release-it/release-it) and some custom scripts.
- Also, now that we no longer need to build on Node 10 or maintain jquery-release scripts, this PR migrates all files in the build folder to use ESM.
- Moved the dist and dist-module wrapper files to `wrappers/` folders to make packaging and cleaning release-only files easier. All references to those files have been updated and it should have no effect on the exports.

*Authors*
- Checking and updating authors has been migrated
  to a custom script in the repo

*Changelog*
- changelogplease is no longer maintained; replace with custom script
- generate changelog in markdown for GitHub releases
- generate changelog in HTML for blog posts

*dist*
- clone dist repo, copy files, and commit/push
- commit tag with dist files on main branch;
  remove dist files from main branch after release

*cdn*
- clone cdn repo, copy files, and commit/push
- create versioned and unversioned copies in cdn/
- generate md5 sums and archives for Google and MSFT

### Checklist ###

* [x] New tests have been added to show the fix or feature works
* [x] If needed, a docs issue/PR was created at https://github.com/jquery/api.jquery.com

This PR currently has feature parity with jquery-release, but these are a few things I'd like to add:

### Future Improvements

- [x] Add workflow to verify new release builds are reproducible
- [x] Document updated release process in `build/release/README.md`
- [x] Generate contributors list with links to GH profiles
- [ ] (stable only) Update version in jquery.com repo
- [ ] (stable only) Update version in jquery-wp-content
- [ ] (stable only) Update version in api.jquery.com demos (entries2html.xsl)
- [ ] (optional) Close GH milestone

### CODE REVIEW

**Comment by timmywil** [comment]
I've tested every step of this. The final test will be to try a real release, but we can wait until after the next beta release.

**Comment by timmywil** [comment]
I realized we can go ahead and review this. I'd like to add a couple things that jquery-release didn't have for stable releases, but those things are optional and could be added in future PRs.

File: build/release/dist.js
Comment (mgol) [inline build/release/dist.js:19]: How are we generating `package.json` now?

File: build/release/post-release.sh
Comment (mgol) [inline build/release/post-release.sh:1]: Let's add some flags here, like `-u`, `-e`, `-o pipefail`?

Comment (mgol) [inline build/release/post-release.sh:1]: This seems like overkill for `sh`, why not:
```suggestion
#!/bin/sh
```
?

**Review by mgol — CHANGES_REQUESTED** [review]
It'll be great to have reproducible builds!

I added some comments.

File: .github/workflows/verify-release.yml
Comment (timmywil) [reply .github/workflows/verify-release.yml:11]: That's not normally how they are numbered. The first beta should have been 4.0.0-beta.0.

File: .gitignore
Comment (timmywil) [reply .gitignore:34]: Everything but changelogs go into the `tmp/` folder. Archives will be under `tmp/archives`. I had almost everything top-level originally, but that was a lot of `.gitignore` entries so I put everything under `tmp/` instead. The changelogs seemed a little different because `changelog.md` is copied to jquery-dist top-level and changelog.html is only for the blog post.

File: .release-it.cjs
Comment (timmywil) [reply .release-it.cjs:1]: Yes, I noted in the PR that I planned on updating the release docs after this PR is merged. I figured I'd update the wiki, but it could be a README in the `build/release`?

File: .github/workflows/verify-release.yml
Comment (mgol) [reply .github/workflows/verify-release.yml:11]: What's the source for this naming strategy?

File: .gitignore
Comment (mgol) [reply .gitignore:34]: Why not put `changelog.html` in `tmp` then?

File: .release-it.cjs
Comment (mgol) [reply .release-it.cjs:1]: A README in `build/release` works for me.

File: build/release/verify.js
Comment (timmywil) [reply build/release/verify.js:79]: no, this works on windows as well

File: .github/workflows/verify-release.yml
Comment (timmywil) [reply .github/workflows/verify-release.yml:11]: Official sources allow for more wiggle room in pre release versions. However, it's moot because this is how release-it works by default and it's not really changeable.

File: .gitignore
Comment (timmywil) [reply .gitignore:34]: I'm not sure that's more helpful. You'd still need the extra `.gitignore` entry and `.*` is even shorter than `.md`.

Comment (timmywil) [reply .gitignore:34]: That said, I could just stick both in `tmp`.

File: .release-it.cjs
Comment (timmywil) [reply .release-it.cjs:28]: Hmm, I didn't think it would work at all without this, but maybe you're right.

File: .release-it.cjs
Comment (timmywil) [reply .release-it.cjs:25]: Yes, that was one of my goals.

File: build/command.js
Comment (timmywil) [reply build/command.js:1]: I also considered that, but `yargs` does provide us more than args parsing. It also gives us a pretty help menu.

File: .release-it.cjs
Comment (timmywil) [reply .release-it.cjs:28]: Will remove.

Comment (timmywil) [reply .release-it.cjs:28]: Wait, I think it's needed the first time though because 4.0.0-beta is detached.

Comment (timmywil) [reply .release-it.cjs:28]: We can remove immediately after.

File: build/release/archive.js
Comment (timmywil) [reply build/release/archive.js:26]: I'm not sure, but let's look into that when we need to support Linux.

File: build/release/cdn.js
Comment (timmywil) [reply build/release/cdn.js:96]: Yes

File: build/release/cdn.js
Comment (timmywil) [reply build/release/cdn.js:113]: This was useful when testing this step on its own. It should be a noop for real releases, but I'd like to leave it.

File: build/release/changelog.js
Comment (timmywil) [reply build/release/changelog.js:13]: I actually wasn't aware we did that. I've never done that. Perhaps it'd make things simpler to stick to one component per commit.

File: build/release/changelog.js
Comment (timmywil) [reply build/release/changelog.js:31]: Is it weird? I thought it was convenient.

File: build/release/changelog.js
Comment (timmywil) [reply build/release/changelog.js:93]: I'm not sure I follow. Does it not already do that? I think the final changelog would compare to the latest RC. But, you bring up a good point that we'll probably want a full 4.0 changelog for the blog post, which we can prepare manually by using `npm run release:changelog -- 3.7.1 4.0.0`.

File: build/release/changelog.js
Comment (timmywil) [reply build/release/changelog.js:85]: I liked it because of how terse it was, but I suppose this could directly affect us if you have another language as your default, like Polish.

File: build/release/changelog.js
Comment (timmywil) [reply build/release/changelog.js:12]: This regex was copied directly from changelogplease, so it's what we had before. I'd say let's keep it the same for now and look into expanding it in a future PR.

File: build/release/changelog.js
Comment (timmywil) [reply build/release/changelog.js:13]: You know, what happens already isn't bad:

```
## CSS

- Tests: Fix tests & support tests under CSS Zoom
```

File: build/release/changelog.js
Comment (timmywil) [reply build/release/changelog.js:101]: Heh, I know. I think that has happened once, but I edited it manually.

File: build/release/dist.js
Comment (timmywil) [reply build/release/dist.js:19]: This is referring to the `package.json` in the `dist` folder, which sets the module type. Everything in the `dist` folder is included, so it's no longer being specified explicitly.

File: build/release/post-release.sh
Comment (timmywil) [reply build/release/post-release.sh:47]: That's part of the regular release process in the [`git` plugin](https://github.com/release-it/release-it/blob/main/docs/git.md)

File: .release-it.cjs
Comment (timmywil) [reply .release-it.cjs:36]: Those are already the defaults, but they can be modified using command line args.

File: build/tasks/build.js
Comment (timmywil) [reply build/tasks/build.js:40]: Yes, it works from any directory as long as you're using npm scripts.

File: build/release/post-release.sh
Comment (timmywil) [reply build/release/post-release.sh:20]: I suppose we could, but with npm scripts as the entry point, the working directory is always the same.

File: build/release/verify.js
Comment (timmywil) [reply build/release/verify.js:40]: Nice catch.

File: build/release/verify.js
Comment (timmywil) [reply build/release/verify.js:125]: I used `.then` to avoid renaming, but I'll just rename.

File: build/release/verify.js
Comment (timmywil) [reply build/release/verify.js:123]: To compare checksums of the tarballs.

File: .gitignore
Comment (timmywil) [reply .gitignore:34]: Actually, could we leave it? I was thinking we could commit the changelog.md to the tag as well, so it would need to stay top-level. And if the `.md` is top-level, it's a little more convenient for me to have the `.html` top-level. I can foresee myself continuing the post release, which removes everything before I've copied the HTML, but that's much less likely if I see the HTML top-level. Yes, I can always generate it again, but I like that it's more in my face.

Comment (mgol) [reply .gitignore:34]: We discussed it on the meeting; we're leaving it as-is.

File: .github/workflows/verify-release.yml
Comment (timmywil) [reply .github/workflows/verify-release.yml:11]: As discussed in the meeting, will update so the next release is `.2`. This may require manual edits to the package.json before initial pre-releases, but we can do that.

File: build/release/README.md
Comment (mgol) [inline build/release/README.md:47]: ...and create one for the next patch release if one already doesn't exist? I guess it will most of the time...

File: build/release/archive.js
Comment (mgol) [reply build/release/archive.js:26]: It will probably be needed when we automate the release via GitHub Actions.

For build breaking Linux would not be acceptable but I guess for a release it is.

File: build/release/README.md
Comment (timmywil) [reply build/release/README.md:47]: I think that's covered in the first step above:

> Ensure all milestoned issues/PRs are closed, or reassign to a new milestone.

Comment (mgol) [reply build/release/README.md:47]: There's an edge case possible where we just fix everything for a release and avoid creating a new one and then there's no milestone ready for new tickets. It's easy to create one when needed, though, so maybe it's good as-is.

File: build/release/changelog.js
Comment (mgol) [reply build/release/changelog.js:31]: It's like using `Array#map` without using its output when you really want `Array#forEach`. It obscures intent - `String#replace` is usually stateless, just like `Array#map`.

Why not use `String#matchAll` followed by `Array#map` with your logic?

File: build/release/README.md
Comment (timmywil) [reply build/release/README.md:47]: I see

File: build/release/changelog.js
Comment (timmywil) [reply build/release/changelog.js:31]: It feels weirder to me to do that because it's looping twice instead of once. Granted, a micro-optimization that doesn't matter for our sample size, but it's not the way I've trained my brain. 😛 I guess this is one of those cases where the ideal doesn't really exist. And yet, if there were some method that was more like `forEach`, the implementation of it might be about identical to `replace`.

File: build/release/README.md
Comment (mgol) [inline build/release/README.md:51]: I’d use `1.` for all points; the diffs are much better when you need to insert something in the middle. Some standards repos are using this strategy, probably for the same reason.

**Review by mgol — APPROVED** [review]
LGTM. Remaining comments can be addressed separately later.

File: build/release/README.md
Comment (timmywil) [reply build/release/README.md:51]: I would normally do that, but with code examples between the numbered list items, the numbers restart. It would work for the first 5 here, but not for any of the below items.

Comment (mgol) [reply build/release/README.md:51]: That just means they aren't formatted properly - so it's even better to `1`-index as then you uncover those bugs immediately. You should indent the code snippets (the triple backticks should also be indented) and they will be pulled into the list.

This should be done as even though the numbers may look correct, we have multiple `<ol>` elements so semantics of the lists are incorrect.

Comment (mgol) [reply build/release/README.md:51]: The same applies to the CDN table at the bottom - since it refers to the 7th point on the list, it should exist within that list.

File: build/release/README.md
Comment (mgol) [inline build/release/README.md:19]: This should also be indented more. Right now, it creates a separate `<ul>` at the same level as the `<ol>` above.

I'd recommend you to look at the rendered HTML and see if the structure makes sense by yourself. You'll uncover all these semantic bugs pretty quickly then.

File: build/release/README.md
Comment (timmywil) [reply build/release/README.md:51]: But indenting the code block ruins the syntax highlighting, at least in vs code preview. I haven't pushed it to github to check there.

Comment (mgol) [reply build/release/README.md:51]: I think semantics are more important - as they affect accessibility - than previews in VS Code. I've used indented code blocks in lists on GitHub and it works perfectly. Maybe the indentation was still not right? I can check later if you don't manage to make it work.

No need to push it to GitHub to test, BTW. You can just put it in a comment, switch to the Preview tab and inspect the HTML.

File: build/release/README.md
Comment (timmywil) [reply build/release/README.md:19]: The extra indent always looked like it registered in code as plain text instead of a list item, but it seems to work.

File: build/release/README.md
Comment (timmywil) [reply build/release/README.md:51]: Take that back. Indenting code blocks or unordered list elements only ruins the markdown code syntax highlighting, and I must have assumed the preview was broken, but it seems to work. However, there are downsides. It reduces the max line length fairly significantly for code blocks.

Comment (timmywil) [reply build/release/README.md]: Seems to be a fair trade-off. I even misnumbered the last item, so this fixes it.

Comment (timmywil) [reply build/release/README.md]: GitHub has condensed any unordered list items and put them on one line. I'm seeing if we can fix that with new lines.

File: build/release/README.md
Comment (timmywil) [reply build/release/README.md]: The extra indent doesn't work on GitHub.

Comment (timmywil) [reply build/release/README.md]: Adding a newline but removing the extra indent seems to work though.

File: build/release/README.md
Comment (timmywil) [reply build/release/README.md]: I don't know if it's actually an accessibility issue, though. The multiple `<ol>` elements were still readable by screen readers. Subsequent ones had a different `start` attribute so the numbers lined up.

**Comment by timmywil** [comment]
Changes since last approval:

1. Now generating a `contributors.html` that includes author names and links to their GH profiles for use in the blog post "Thanks" section. So much better than what I was using before!
2. Added a "Setup" section to the release README, which includes basic setup instructions like cloning and installing dependencies, but more importantly setup instructions for logging into npm and ensuring a github token is set in the environment, with an easy link to create a new token if necessary.
3. Updated release-it to 17.5.0, which includes the new `--preReleaseBase` option, which should be set to `1` for the first pre-release of a kind (e.g. `--preRelease=beta --preReleaseBase=1` to release `.beta.1` instead of `.beta.0`). This has also been documented.
4. Updated AUTHORS.txt to prepare for the upcoming release.

File: build/release/README.md
Comment (mgol) [inline build/release/README.md:110]: These two should also be `1.`s.

**Review by mgol — APPROVED** [review]
Still LGTM after the latest changes. Just a small note on leftover larger numbers in lists.

**Pull request merged**
